src/alignment/sparse.rs

```diff
@@ -0,0 +1,218 @@
+// Copyright 2014-2015 Patrick Marks
+// Licensed under the MIT license (http://opensource.org/licenses/MIT)
+// This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Calculate 'sparse' alignments from kmer matches. Can be much faster than 
+//! Smith-Waterman for long string, when a large enough k is used. 
+//! Complexity: O(n * log(n)) for a pair of strings with n k-kmer matches. This
+//! approach is useful for generating an approximate 'backbone' alignments 
+//! between two long sequences, for example in long-read alignment or 
+//! genome-genome alignment. The backbone alignment can be used as-is, or can serve
+//! as a guide for a banded alignment.  By tuning k so that len(query) + len(reference) < 4^k,
+//! the number of false positive kmer matches is kept small, resulting in very
+//! fast run times for long strings.
+//!
+//! # Example
+//!
+//! ```
+//! use bio::alignment::sparse::*;
+//!
+//! let s1 =   "ACGTACGATAGGTA";
+//! let s2 = "TTACGTACGATAGGTATT";
+//! let k = 8;
+//! let matches = find_kmer_matches(&s1, &s2, k);
+//! let (tb, score,  _) = lcskpp(&matches, k);
+//! let match_path: Vec<(u32,u32)> = tb.iter().map(|i| matches[*i]).collect();
+//! assert_eq!(match_path, vec![(0,2), (1,3), (2,4), (3,5), (4,6), (5,7), (6,8)]);
+//! assert_eq!(score, 14);
+
+use std::cmp::max;
+use data_structures::bit_tree::MaxBitTree;
+use data_structures::qgram_index::QGramIndex;
+use alphabets::Alphabet;
+
+/// Sparse DP routine for Longest Common Subsequence in length k substrings.  Also known of LCSk++
+/// From LCSk++: Practical similarity metric for long strings. Filip Pavetić, Goran Žužić, Mile Šikić
+/// Paper here :https://arxiv.org/abs/1407.2407.  Original implementation here: https://github.com/fpavetic/lcskpp
+/// matches is a tuple of the positions of all k-mer matches between two strings. matches vector must be sorted by caller.
+/// The first element of the return tuple is the LCSk++ path, represented as vector of indices into the matches vector.
+/// The second element of the return tuple is the score of the path, which is the number of bases covered by the matched kmers.
+/// The third element of the return tuple is full DP vector, which can generally be ignored. (It may be useful for testing purposes).
```

Comment:
The matches vector should be explained further. What is the meaning of the two numbers (I guess start positions of the kmer in the first and the second sequence?